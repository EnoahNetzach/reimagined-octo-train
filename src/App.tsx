import { createBrowserRouter, RouterProvider } from 'react-router-dom'
import { ApolloClient, InMemoryCache, ApolloProvider } from '@apollo/client'
import Character from './Character.tsx'
import ErrorPage from './ErrorPage'
import MainLayout from './MainLayout.tsx'

const router = createBrowserRouter([
  {
    path: '/',
    element: <MainLayout />,
    errorElement: <ErrorPage />,
    children: [{ path: '/character/:characterId', element: <Character />, errorElement: <ErrorPage /> }],
  },
])

const client = new ApolloClient({
  uri: 'https://rickandmortyapi.com/graphql',
  cache: new InMemoryCache({
    typePolicies: {
      Query: {
        fields: {
          characters: {
            keyArgs: ['filter'],
            merge: (existing = {}, incoming) => ({
              ...existing,
              ...incoming,
              results: [...(existing.results ?? []), ...(incoming.results ?? [])],
            }),
          },
        },
      },
    },
  }),
})

export default function App() {
  return (
    <ApolloProvider client={client}>
      <RouterProvider router={router} />
    </ApolloProvider>
  )
}
