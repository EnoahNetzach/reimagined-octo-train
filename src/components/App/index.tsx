import { Suspense } from 'react'
import { createBrowserRouter, RouterProvider } from 'react-router-dom'
import { ApolloClient, InMemoryCache, ApolloProvider } from '@apollo/client'
import { BatchHttpLink } from '@apollo/client/link/batch-http'
import { RetryLink } from '@apollo/client/link/retry'
import Character from '~/components/Character'
import ErrorPage from '~/components/ErrorPage'
import MainLayout from '~/components/MainLayout'
import { basePath } from '~/configs'

const router = createBrowserRouter(
  [
    {
      path: '/',
      element: <MainLayout />,
      errorElement: <ErrorPage />,
      children: [{ path: '/character/:characterId', element: <Character />, errorElement: <ErrorPage /> }],
    },
  ],
  {
    basename: basePath,
  },
)

const client = new ApolloClient({
  cache: new InMemoryCache({
    typePolicies: {
      Query: {
        fields: {
          characters: {
            keyArgs: ['filter'],
            merge: (existing = {}, incoming) => ({
              ...existing,
              ...incoming,
              results: [...(existing.results ?? []), ...(incoming.results ?? [])],
            }),
          },
        },
      },
    },
  }),
  link: new RetryLink({ attempts: { max: Infinity }, delay: { initial: 100 } }).concat(
    new BatchHttpLink({ uri: 'https://rickandmortyapi.com/graphql' }),
  ),
})

export default function App() {
  return (
    <Suspense fallback={null}>
      <ApolloProvider client={client}>
        <RouterProvider router={router} />
      </ApolloProvider>
    </Suspense>
  )
}
