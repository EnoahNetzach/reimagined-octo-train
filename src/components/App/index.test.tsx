import { PropsWithChildren } from 'react'
import { createRenderer } from 'react-test-renderer/shallow'
import App from '.'

jest.mock('@apollo/client', () => ({
  ApolloClient: class {
    options: unknown
    constructor(options: unknown) {
      this.options = options
    }
  },
  InMemoryCache: class {
    options: unknown
    constructor(options: unknown) {
      this.options = options
    }
  },
  ApolloProvider: ({ children }: PropsWithChildren) => <>{children}</>,
}))

jest.mock('react-router-dom', () => ({
  createBrowserRouter: (routes: unknown[]) => routes,
  RouterProvider: ({ children }: PropsWithChildren) => <>{children}</>,
}))

jest.mock(
  '~/components/Character',
  () =>
    function Character() {
      return <>Character Mock</>
    },
)

jest.mock(
  '~/components/ErrorPage',
  () =>
    function ErrorPage() {
      return <>ErrorPage Mock</>
    },
)

jest.mock(
  '~/components/MainLayout',
  () =>
    function MainLayout() {
      return <>MainLayout Mock</>
    },
)

describe('App', () => {
  it('mounts', () => {
    const renderer = createRenderer()
    renderer.render(<App />)
    expect(renderer.getRenderOutput()).toMatchSnapshot()
  })
})
