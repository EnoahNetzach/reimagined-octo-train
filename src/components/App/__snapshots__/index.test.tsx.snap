// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App mounts 1`] = `
<React.Suspense
  fallback={null}
>
  <ApolloProvider
    client={
      ApolloClient {
        "options": {
          "cache": InMemoryCache {
            "options": {
              "typePolicies": {
                "Query": {
                  "fields": {
                    "characters": {
                      "keyArgs": [
                        "filter",
                      ],
                      "merge": [Function],
                    },
                  },
                },
              },
            },
          },
          "uri": "https://rickandmortyapi.com/graphql",
        },
      }
    }
  >
    <RouterProvider
      router={
        [
          {
            "children": [
              {
                "element": <Character />,
                "errorElement": <ErrorPage />,
                "path": "/character/:characterId",
              },
            ],
            "element": <MainLayout />,
            "errorElement": <ErrorPage />,
            "path": "/",
          },
        ]
      }
    />
  </ApolloProvider>
</React.Suspense>
`;
